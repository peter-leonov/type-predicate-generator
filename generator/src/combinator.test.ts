import { expect, describe, it } from "vitest";
import {
  array,
  combineInvalid,
  combineValid,
  modelToCombinator,
  object,
  union,
  value,
  ValueGenerator,
  type Value,
} from "./combinator";
import {
  AliasType,
  ArrayType,
  LiteralType,
  ObjectType,
  PrimitiveType,
  UnionType,
} from "./model";

const stringified = new Map<unknown, string>();
function stringifyCombination(value: unknown): string {
  const cached = stringified.get(value);
  if (cached != undefined) {
    return cached;
  }
  const res = JSON.stringify(value, (_, v) => {
    return typeof v == "symbol" ? String(v) : v;
  });
  stringified.set(value, res);
  return res;
}

function compareCombinations(a: unknown, b: unknown) {
  return stringifyCombination(a) < stringifyCombination(b) ? -1 : 0;
}

function sort(combinations: unknown[]) {
  return combinations.sort(compareCombinations);
}

function valid(v: ValueGenerator<Value>) {
  return sort(combineValid(v));
}

function invalid(v: ValueGenerator<Value>) {
  return sort(combineInvalid(v));
}

for (const [name, f] of [
  ["valid", valid] as const,
  ["invalid", invalid] as const,
]) {
  describe(name, () => {
    it("all at once", () => {
      const obj = object<Value>({
        a: union([value(1), value(2)]),
        b: union([value(true), value(false)]),
        c: array(union([value("a"), value("b")])),
        d: object({
          d2: union([value(null), value(undefined)]),
        }),
      });

      expect(f(obj)).toMatchSnapshot();
    });

    describe("value", () => {
      it("number", () => {
        expect(f(value(1))).toMatchSnapshot();
      });
      it("string", () => {
        expect(f(value("a"))).toMatchSnapshot();
      });
      it("null", () => {
        expect(f(value(null))).toMatchSnapshot();
      });
    });

    describe("union", () => {
      it("of a single value", () => {
        expect(f(union([value(1)]))).toMatchSnapshot();
      });

      it("of several values", () => {
        expect(
          f(union([value(1), value(2), value(3)]))
        ).toMatchSnapshot();
      });

      it("of a union", () => {
        expect(
          f(union([union([value(1), value(2), value(3)])]))
        ).toMatchSnapshot();
      });

      it("of unions", () => {
        expect(
          f(
            union([
              union([value(1), value(2)]),
              union([value(3), value(4)]),
            ])
          )
        ).toMatchSnapshot();
      });
    });

    describe("array", () => {
      it("of a single value", () => {
        expect(f(array(value(1)))).toMatchSnapshot();
      });

      it("of a union", () => {
        expect(
          f(array(union([value(1), value(2)])))
        ).toMatchSnapshot();
      });

      it("of an array of an array of a value", () => {
        expect(f(array(array(array(value(1)))))).toMatchSnapshot();
      });

      it("of an array of an array of a union", () => {
        expect(
          f(array(array(array(union([value(1), value(2)])))))
        ).toMatchSnapshot();
      });
    });

    describe("object", () => {
      it("empty", () => {
        expect(f(object({}))).toMatchSnapshot();
      });

      it("of a single value property", () => {
        expect(f(object({ a: value("A") }))).toMatchSnapshot();
      });

      it("of a several value properties", () => {
        expect(
          f(object({ a: value("A"), b: value("B"), c: value("C") }))
        ).toMatchSnapshot();
      });

      it("of a single union property", () => {
        expect(
          f(object({ a: union([value("A1"), value("A2")]) }))
        ).toMatchSnapshot();
      });

      it("of a several union properties", () => {
        expect(
          f(
            object({
              a: union([value("A1")]),
              b: union([value("B1"), value("B2")]),
              c: union([value("C1"), value("C2"), value("C3")]),
            })
          )
        ).toMatchSnapshot();
      });

      it("of a several objects with value properties", () => {
        expect(
          f(
            object({
              a: object({ a1: value("A1"), a2: value("A2") }),
              b: object({ b1: value("B1"), b2: value("B2") }),
              c: object({ c1: value("C1"), c2: value("C2") }),
            })
          )
        ).toMatchSnapshot();
      });

      it("of a several objects with union properties", () => {
        expect(
          f(
            object({
              a: object({ aa: union([value("AA1"), value("AA2")]) }),
              b: object({ bb: union([value("BB1"), value("BB2")]) }),
              c: object({ cc: union([value("CC1"), value("CC2")]) }),
            })
          )
        ).toMatchSnapshot();
      });

      it("of a several nested objects with a value property", () => {
        expect(
          f(
            object({
              a: object({
                b: object({ c: object({ d: value("D") }) }),
              }),
            })
          )
        ).toMatchSnapshot();
      });

      it("of a several nested objects with a union property", () => {
        expect(
          f(
            object({
              a: object({
                b: object({
                  c: object({ d: union([value("D1"), value("D2")]) }),
                }),
              }),
            })
          )
        ).toMatchSnapshot();
      });
    });
  });
}

describe(modelToCombinator.name, () => {
  describe("LiteralType", () => {
    it("undefined", () => {
      expect(
        combineValid(
          modelToCombinator(new LiteralType({}, undefined))
        )
      ).toMatchInlineSnapshot(`
        [
          undefined,
        ]
      `);
    });
    it("null", () => {
      expect(
        combineValid(modelToCombinator(new LiteralType({}, null)))
      ).toMatchInlineSnapshot(`
        [
          null,
        ]
      `);
    });
    it("string", () => {
      expect(
        combineValid(modelToCombinator(new LiteralType({}, "foo")))
      ).toMatchInlineSnapshot(`
        [
          "foo",
        ]
      `);
    });
    it("number", () => {
      expect(combineValid(modelToCombinator(new LiteralType({}, 42))))
        .toMatchInlineSnapshot(`
        [
          42,
        ]
      `);
    });
    it("boolean", () => {
      expect(
        combineValid(modelToCombinator(new LiteralType({}, true)))
      ).toMatchInlineSnapshot(`
        [
          true,
        ]
      `);
      expect(
        combineValid(modelToCombinator(new LiteralType({}, false)))
      ).toMatchInlineSnapshot(`
        [
          false,
        ]
      `);
    });
  });

  describe("PrimitiveType", () => {
    it("string", () => {
      expect(
        combineValid(
          modelToCombinator(new PrimitiveType({}, "string"))
        )
      ).toMatchInlineSnapshot(`
        [
          "string",
        ]
      `);
    });
    it("number", () => {
      expect(
        combineValid(
          modelToCombinator(new PrimitiveType({}, "number"))
        )
      ).toMatchInlineSnapshot(`
        [
          0,
          42,
        ]
      `);
    });
    it("boolean", () => {
      expect(
        combineValid(
          modelToCombinator(new PrimitiveType({}, "boolean"))
        )
      ).toMatchInlineSnapshot(`
        [
          true,
          false,
        ]
      `);
    });
  });

  describe("ObjectType", () => {
    it("empty", () => {
      expect(combineValid(modelToCombinator(new ObjectType({}, {}))))
        .toMatchInlineSnapshot(`
          [
            {},
          ]
        `);
    });
    it("nested", () => {
      expect(
        combineValid(
          modelToCombinator(
            new ObjectType({}, { a: new ObjectType({}, {}) })
          )
        )
      ).toMatchInlineSnapshot(`
        [
          {
            "a": {},
          },
        ]
      `);
    });
    it("literal attributes", () => {
      expect(
        combineValid(
          modelToCombinator(
            new ObjectType(
              {},
              {
                a: new LiteralType({}, 1),
                b: new LiteralType({}, "foo"),
                c: new LiteralType({}, true),
              }
            )
          )
        )
      ).toMatchInlineSnapshot(`
        [
          {
            "a": 1,
            "b": "foo",
            "c": true,
          },
        ]
      `);
    });
  });

  describe("UnionType", () => {
    it("empty", () => {
      expect(
        combineValid(modelToCombinator(new UnionType({}, [])))
      ).toMatchInlineSnapshot(`[]`);
    });
    it("of literal types", () => {
      expect(
        combineValid(
          modelToCombinator(
            new UnionType({}, [
              new LiteralType({}, 1),
              new LiteralType({}, 2),
              new LiteralType({}, 3),
            ])
          )
        )
      ).toMatchInlineSnapshot(`
        [
          1,
          2,
          3,
        ]
      `);
    });
  });

  describe("ArrayType", () => {
    it("empty", () => {
      expect(
        combineValid(
          modelToCombinator(
            new ArrayType({}, new PrimitiveType({}, "boolean"))
          )
        )
      ).toMatchInlineSnapshot(`
        [
          [],
          [
            true,
          ],
          [
            false,
          ],
        ]
      `);
    });
  });

  describe("AliasType", () => {
    it("empty", () => {
      expect(() =>
        combineValid(modelToCombinator(new AliasType({}, "Foo")))
      ).toThrow();
    });
  });
});
