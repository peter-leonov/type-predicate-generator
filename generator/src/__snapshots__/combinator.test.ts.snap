// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`invalid > array > of an array of an array of a union 1`] = `
[
  Symbol(invalidValue),
  [
    Symbol(invalidValue),
  ],
  [
    [
      Symbol(invalidValue),
    ],
  ],
  [
    [
      [
        Symbol(invalidValue),
      ],
    ],
  ],
  [
    [
      [
        Symbol(invalidValue),
      ],
    ],
  ],
]
`;

exports[`invalid > array > of an array of an array of a value 1`] = `
[
  Symbol(invalidValue),
  [
    Symbol(invalidValue),
  ],
  [
    [
      Symbol(invalidValue),
    ],
  ],
  [
    [
      [
        Symbol(invalidValue),
      ],
    ],
  ],
]
`;

exports[`invalid > array > single value 1`] = `
[
  Symbol(invalidValue),
  [
    Symbol(invalidValue),
  ],
]
`;

exports[`invalid > object > empty 1`] = `
[
  Symbol(invalidValue),
]
`;

exports[`invalid > object > of a several value properties 1`] = `
[
  Symbol(invalidValue),
  {
    "b": "B",
    "c": "C",
  },
  {
    "a": Symbol(invalidValue),
    "b": "B",
    "c": "C",
  },
  {
    "a": "A",
    "c": "C",
  },
  {
    "a": "A",
    "b": Symbol(invalidValue),
    "c": "C",
  },
  {
    "a": "A",
    "b": "B",
  },
  {
    "a": "A",
    "b": "B",
    "c": Symbol(invalidValue),
  },
]
`;

exports[`invalid > object > of a single union property 1`] = `
[
  Symbol(invalidValue),
  {},
  {
    "a": Symbol(invalidValue),
  },
  {
    "a": Symbol(invalidValue),
  },
]
`;

exports[`invalid > object > of a single value property 1`] = `
[
  Symbol(invalidValue),
  {},
  {
    "a": Symbol(invalidValue),
  },
]
`;

exports[`invalid > union > couple of values 1`] = `
[
  Symbol(invalidValue),
  Symbol(invalidValue),
  Symbol(invalidValue),
]
`;

exports[`invalid > union > single value 1`] = `
[
  Symbol(invalidValue),
]
`;

exports[`invalid > value > number 1`] = `
[
  Symbol(invalidValue),
]
`;

exports[`valid > all at once 1`] = `
[
  {
    "a": 1,
    "b": true,
    "c": [
      "a",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 1,
    "b": true,
    "c": [
      "a",
    ],
    "d": {
      "d2": undefined,
    },
  },
  {
    "a": 1,
    "b": true,
    "c": [
      "b",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 1,
    "b": true,
    "c": [
      "b",
    ],
    "d": {
      "d2": undefined,
    },
  },
  {
    "a": 1,
    "b": false,
    "c": [
      "a",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 1,
    "b": false,
    "c": [
      "a",
    ],
    "d": {
      "d2": undefined,
    },
  },
  {
    "a": 1,
    "b": false,
    "c": [
      "b",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 1,
    "b": false,
    "c": [
      "b",
    ],
    "d": {
      "d2": undefined,
    },
  },
  {
    "a": 2,
    "b": true,
    "c": [
      "a",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 2,
    "b": true,
    "c": [
      "a",
    ],
    "d": {
      "d2": undefined,
    },
  },
  {
    "a": 2,
    "b": true,
    "c": [
      "b",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 2,
    "b": true,
    "c": [
      "b",
    ],
    "d": {
      "d2": undefined,
    },
  },
  {
    "a": 2,
    "b": false,
    "c": [
      "a",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 2,
    "b": false,
    "c": [
      "a",
    ],
    "d": {
      "d2": undefined,
    },
  },
  {
    "a": 2,
    "b": false,
    "c": [
      "b",
    ],
    "d": {
      "d2": null,
    },
  },
  {
    "a": 2,
    "b": false,
    "c": [
      "b",
    ],
    "d": {
      "d2": undefined,
    },
  },
]
`;

exports[`valid > array > of a single value 1`] = `
[
  [
    1,
  ],
]
`;

exports[`valid > array > of a union 1`] = `
[
  [
    1,
  ],
  [
    2,
  ],
]
`;

exports[`valid > array > of an array of an array of a union 1`] = `
[
  [
    [
      [
        1,
      ],
    ],
  ],
  [
    [
      [
        2,
      ],
    ],
  ],
]
`;

exports[`valid > array > of an array of an array of a value 1`] = `
[
  [
    [
      [
        1,
      ],
    ],
  ],
]
`;

exports[`valid > object > empty 1`] = `
[
  {},
]
`;

exports[`valid > object > of a several nested objects with a union property 1`] = `
[
  {
    "a": {
      "b": {
        "c": {
          "d": "D1",
        },
      },
    },
  },
  {
    "a": {
      "b": {
        "c": {
          "d": "D2",
        },
      },
    },
  },
]
`;

exports[`valid > object > of a several nested objects with a value property 1`] = `
[
  {
    "a": {
      "b": {
        "c": {
          "d": "D",
        },
      },
    },
  },
]
`;

exports[`valid > object > of a several objects with union properties 1`] = `
[
  {
    "a": {
      "aa": "AA1",
    },
    "b": {
      "bb": "BB1",
    },
    "c": {
      "cc": "CC1",
    },
  },
  {
    "a": {
      "aa": "AA1",
    },
    "b": {
      "bb": "BB1",
    },
    "c": {
      "cc": "CC2",
    },
  },
  {
    "a": {
      "aa": "AA1",
    },
    "b": {
      "bb": "BB2",
    },
    "c": {
      "cc": "CC1",
    },
  },
  {
    "a": {
      "aa": "AA1",
    },
    "b": {
      "bb": "BB2",
    },
    "c": {
      "cc": "CC2",
    },
  },
  {
    "a": {
      "aa": "AA2",
    },
    "b": {
      "bb": "BB1",
    },
    "c": {
      "cc": "CC1",
    },
  },
  {
    "a": {
      "aa": "AA2",
    },
    "b": {
      "bb": "BB1",
    },
    "c": {
      "cc": "CC2",
    },
  },
  {
    "a": {
      "aa": "AA2",
    },
    "b": {
      "bb": "BB2",
    },
    "c": {
      "cc": "CC1",
    },
  },
  {
    "a": {
      "aa": "AA2",
    },
    "b": {
      "bb": "BB2",
    },
    "c": {
      "cc": "CC2",
    },
  },
]
`;

exports[`valid > object > of a several objects with value properties 1`] = `
[
  {
    "a": {
      "a1": "A1",
      "a2": "A2",
    },
    "b": {
      "b1": "B1",
      "b2": "B2",
    },
    "c": {
      "c1": "C1",
      "c2": "C2",
    },
  },
]
`;

exports[`valid > object > of a several union properties 1`] = `
[
  {
    "a": "A1",
    "b": "B1",
    "c": "C1",
  },
  {
    "a": "A1",
    "b": "B1",
    "c": "C2",
  },
  {
    "a": "A1",
    "b": "B1",
    "c": "C3",
  },
  {
    "a": "A1",
    "b": "B2",
    "c": "C1",
  },
  {
    "a": "A1",
    "b": "B2",
    "c": "C2",
  },
  {
    "a": "A1",
    "b": "B2",
    "c": "C3",
  },
]
`;

exports[`valid > object > of a several value properties 1`] = `
[
  {
    "a": "A",
    "b": "B",
    "c": "C",
  },
]
`;

exports[`valid > object > of a single union property 1`] = `
[
  {
    "a": "A1",
  },
  {
    "a": "A2",
  },
]
`;

exports[`valid > object > of a single value property 1`] = `
[
  {
    "a": "A",
  },
]
`;

exports[`valid > union > of a single value 1`] = `
[
  1,
]
`;

exports[`valid > union > of a union 1`] = `
[
  1,
  2,
  3,
]
`;

exports[`valid > union > of several values 1`] = `
[
  1,
  2,
  3,
]
`;

exports[`valid > union > of unions 1`] = `
[
  1,
  2,
  3,
  4,
]
`;

exports[`valid > value > null 1`] = `
[
  null,
]
`;

exports[`valid > value > number 1`] = `
[
  1,
]
`;

exports[`valid > value > string 1`] = `
[
  "a",
]
`;
